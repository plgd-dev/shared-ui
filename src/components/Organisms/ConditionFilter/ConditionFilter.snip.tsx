import React, { useMemo, useState } from 'react'
import { OptionType } from '../../Atomic/FormSelect/FormSelect.types'
import ConditionFilter from './ConditionFilter'
import StatusTag from '../../Atomic/StatusTag'
import FormGroup from '../../Atomic/FormGroup'
import FormLabel from '../../Atomic/FormLabel'
import FormSelect from '../../Atomic/FormSelect'
import FormInput from '../../Atomic/FormInput'
import Button from '../../Atomic/Button'
import IconPlus from '../../Atomic/Icon/components/IconPlus'

type Props = {
    dataTestId?: string
    isTest?: boolean
}

export const FormSelectConditionFilter = (props: Props) => {
    const { dataTestId, isTest } = props
    const [items, setItems] = useState<string[]>([])

    const [options, setOptions] = useState<OptionType[]>([
        { value: 'device-1', label: 'Device 1' },
        { value: 'device-2', label: 'Device 2' },
        { value: 'device-3', label: 'Device 3' },
        { value: 'device-4', label: 'Device 4' },
        { value: 'device-5', label: 'Device 5' },
    ])

    const value = useMemo(() => items.map((id: string) => options?.find((o) => o.value === id) || { value: id, label: id }), [items, options])

    return (
        <ConditionFilter
            dataTestId={dataTestId}
            isTest={isTest}
            listName='List of Selected Item'
            listOfItems={items.map((id) => options?.find((o) => o.value === id)?.label || id)}
            onItemDelete={(key) => {
                const newItems = items.filter((_, i) => i !== key)
                setItems(newItems)
            }}
            status={
                <StatusTag lowercase={false} variant={items.length > 0 ? 'success' : 'normal'}>
                    {items.length > 0 ? 'Set up' : 'Not set'}
                </StatusTag>
            }
            title='FormSelect'
        >
            <FormGroup id='items'>
                <FormLabel text='Select Item' />
                <FormSelect
                    checkboxOptions
                    creatable
                    isMulti
                    footerLinksLeft={[
                        {
                            title: 'Reset',
                            onClick: () => {
                                setItems([])
                            },
                        },
                        {
                            title: 'Done',
                            variant: 'primary',
                            onClick: (values: OptionType[]) => {
                                const rawValue = values.map((v) => v.value.toString())
                                setItems(rawValue)
                            },
                        },
                    ]}
                    i18n={{
                        itemSelected: 'Item selected',
                        itemsSelected: 'Items selected',
                    }}
                    menuPortalTarget={document.body}
                    menuZIndex={100}
                    name='items'
                    onChange={(values: OptionType[]) => {
                        const rawValue = values.map((v) => v.value.toString())
                        setItems(rawValue)
                    }}
                    onCreateOption={(value: string | number) => {
                        setOptions((prev) => [...prev, { value: value.toString(), label: value.toString() }])
                    }}
                    options={options}
                    placeholder='Select or Create'
                    value={value}
                />
            </FormGroup>
        </ConditionFilter>
    )
}

export const FormInputConditionFilter = () => {
    const [items, setItems] = useState<string[]>([])
    const [value, setValue] = useState<string>('')

    return (
        <ConditionFilter
            listName='List of selected Items'
            listOfItems={items}
            onItemDelete={(key) => {
                const newVal = items?.filter((_, i) => i !== key)
                setItems(newVal)
            }}
            status={
                <StatusTag lowercase={false} variant={items?.length > 0 ? 'success' : 'normal'}>
                    {items?.length > 0 ? 'Set up' : 'Not set'}
                </StatusTag>
            }
            title='FormInput'
        >
            <div style={{ display: 'flex', alignItems: 'flex-end', gap: '8px' }}>
                <FormGroup id='resourceTypeFilter' marginBottom={false} style={{ flex: '1 1 auto' }}>
                    <FormLabel text='Add manual data' />
                    <FormInput
                        compactFormComponentsView={false}
                        onChange={(e) => setValue(e.target.value)}
                        onKeyPress={(e) => {
                            if (e.key === 'Enter' && e.target.value !== '') {
                                e.preventDefault()
                                const newVal = [...items, e.target.value]
                                setItems(newVal)
                                setValue('')
                            }
                        }}
                        value={value}
                    />
                </FormGroup>
                <Button
                    disabled={value === ''}
                    icon={<IconPlus />}
                    onClick={() => {
                        const newVal = [...items, value]
                        setItems(newVal)
                        setValue('')
                    }}
                    size='small'
                    style={{
                        position: 'relative',
                        bottom: '2px',
                    }}
                    variant='secondary'
                >
                    Add
                </Button>
            </div>
        </ConditionFilter>
    )
}

export const FormInputSingleConditionFilter = () => {
    const [value, setValue] = useState('')
    return (
        <ConditionFilter
            listName='Selected'
            status={
                <StatusTag lowercase={false} variant={value !== '' ? 'success' : 'normal'}>
                    {value !== '' ? 'Set up' : 'Not set'}
                </StatusTag>
            }
            title='FormInput single'
        >
            <FormGroup id='jqExpressionFilter'>
                <FormLabel text='Add manual data' />
                <FormInput
                    compactFormComponentsView={false}
                    onChange={(e) => {
                        setValue(e.target.value)
                    }}
                    onKeyPress={(e) => {
                        if (e.key === 'Enter') {
                            e.preventDefault()
                            setValue(e.target.value)
                        }
                    }}
                    value={value}
                />
            </FormGroup>
        </ConditionFilter>
    )
}
