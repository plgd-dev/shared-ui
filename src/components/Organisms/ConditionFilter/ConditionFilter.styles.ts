import { css } from '@emotion/react'
import { getThemeColor, ThemeType } from '../../Atomic/_theme'
import { fontSecondary } from '../../Atomic/_utils/commonStyles'

export const filter = (theme: ThemeType) => css`
    border-radius: 8px;
    border: 1px solid ${getThemeColor(theme, 'ConditionFilter.border')};
    overflow: hidden;
`

export const header = (theme: ThemeType) => css`
    padding: 16px 24px;
    display: flex;
    align-items: center;
    justify-content: space-between;
    cursor: pointer;
    background: ${getThemeColor(theme, 'ConditionFilter.header.background')};

    &:hover {
        background: ${getThemeColor(theme, 'ConditionFilter.header.hover.background')};

        .expander {
            color: ${getThemeColor(theme, 'ConditionFilter.header.hover.expander.color')};
        }
    }
`

const headlineBase = css`
    font: ${fontSecondary};
    font-size: 16px;
    font-style: normal;
    font-weight: 700;
    line-height: 150%;
    letter-spacing: -0.5px;
`

export const title = (theme: ThemeType) => css`
    ${headlineBase};
    color: ${getThemeColor(theme, 'ConditionFilter.title.color')};
    display: flex;
    align-items: center;
`

export const left = css`
    display: flex;
    align-items: center;
    gap: 8px;
`

export const right = css`
    flex: 0 0 32px;
`

export const expander = (theme: ThemeType) => css`
    color: ${getThemeColor(theme, 'ConditionFilter.expander.color')};
    background: ${getThemeColor(theme, 'ConditionFilter.expander.background')};
    padding: 6px;
    border-radius: 8px;
    transition: all 0.3s;
    width: 32px;
    height: 32px;

    &:hover {
        color: ${getThemeColor(theme, 'ConditionFilter.expander.hover.color')};
    }
`

export const content = (theme: ThemeType) => css`
    padding: 24px;
    border-top: 1px solid ${getThemeColor(theme, 'ConditionFilter.border')};
`

export const inputBox = (theme: ThemeType) => css`
    padding: 16px 24px 24px 24px;
    flex-direction: column;
    align-items: flex-start;
    gap: 10px;
    border-radius: 8px;
    background: ${getThemeColor(theme, 'ConditionFilter.inputBox.background')};
`

export const list = css`
    padding: 0;
    margin: 0;
    list-style: none;
`

export const listItem = (theme: ThemeType) => css`
    color: ${getThemeColor(theme, 'ConditionFilter.list.item.color')};
`

export const listIcon = (theme: ThemeType) => css`
    color: ${getThemeColor(theme, 'ConditionFilter.list.icon.color')};
    cursor: pointer;
    transition: all 0.3s;

    &:hover {
        color: ${getThemeColor(theme, 'ConditionFilter.list.icon.hover.color')};
    }
`
