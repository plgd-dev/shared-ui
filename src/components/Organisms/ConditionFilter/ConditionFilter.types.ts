import { ReactNode } from 'react'

export type Props = {
    className?: string
    children?: ReactNode
    dataTestId?: string
    defaultOpen?: boolean
    id?: string
    isTest?: boolean
    listName?: string
    listOfItems?: string[]
    onItemDelete?: (item: number) => void
    status?: ReactNode
    title: string
}
