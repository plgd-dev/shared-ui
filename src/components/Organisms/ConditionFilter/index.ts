export { default } from './ConditionFilter'
export * from './ConditionFilter'
