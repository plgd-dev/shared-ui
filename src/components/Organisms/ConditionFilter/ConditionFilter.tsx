import React, { FC, useState } from 'react'
import { AnimatePresence, motion } from 'framer-motion'

import { Props } from './ConditionFilter.types'
import * as styles from './ConditionFilter.styles'
import IconArrowDownNoPadding from '../../Atomic/Icon/components/IconArrowDownNoPadding'
import { convertSize } from '../../Atomic/Icon'
import Spacer from '../../Atomic/Spacer'
import Headline from '../../Atomic/Headline'
import SimpleStripTable from '../../Atomic/SimpleStripTable'
import IconTrash from '../../Atomic/Icon/components/IconTrash'
import isFunction from 'lodash/isFunction'

const ConditionFilter: FC<Props> = (props) => {
    const { className, children, dataTestId, defaultOpen, listName, listOfItems, onItemDelete, status, title } = props

    const [show, setShow] = useState(defaultOpen ?? false)

    return (
        <div className={className} css={styles.filter} data-test-id={dataTestId}>
            <div css={styles.header} onClick={() => setShow(!show)}>
                <div css={styles.left}>
                    <div css={styles.title}>{title}</div>
                    {status}
                </div>
                <div css={styles.right}>
                    <motion.div
                        animate={{
                            rotate: show ? 180 : 0,
                        }}
                        className='expander'
                        css={styles.expander}
                        transition={{
                            duration: 0.3,
                        }}
                    >
                        <IconArrowDownNoPadding {...convertSize(20)} />
                    </motion.div>
                </div>
            </div>
            <AnimatePresence>
                {show && (
                    <motion.div
                        animate={{ opacity: 1, height: 'auto' }}
                        exit={{
                            opacity: 0,
                            height: 0,
                        }}
                        initial={{ opacity: 0, height: 0 }}
                        transition={{
                            duration: 0.3,
                        }}
                    >
                        <div css={styles.content}>
                            <div css={styles.inputBox}>{children}</div>

                            {listOfItems && listOfItems.length > 0 && listName && isFunction(onItemDelete) && (
                                <Spacer type='pt-6'>
                                    <Spacer type='mb-3'>
                                        <Headline type='h6'>{listName}</Headline>
                                    </Spacer>

                                    <SimpleStripTable
                                        lastRowBorder={false}
                                        leftColSize={10}
                                        rightColSize={2}
                                        rows={listOfItems.map((item, key) => ({
                                            attribute: <span css={styles.listItem}>{item}</span>,
                                            value: <IconTrash css={styles.listIcon} onClick={isFunction(onItemDelete) ? () => onItemDelete(key) : undefined} />,
                                        }))}
                                    />
                                </Spacer>
                            )}
                        </div>
                    </motion.div>
                )}
            </AnimatePresence>
        </div>
    )
}

ConditionFilter.displayName = 'ConditionFilter'

export default ConditionFilter
