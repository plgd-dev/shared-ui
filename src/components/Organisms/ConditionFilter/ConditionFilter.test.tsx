import React from 'react'
import { fireEvent, waitFor } from '@testing-library/react'

import { render } from '../../../../test/jest-wrapper'
import { FormInputConditionFilter, FormInputSingleConditionFilter, FormSelectConditionFilter } from './ConditionFilter.snip'
import { forEach } from 'lodash'

describe('<ConditionFilter>', () => {
    it('render correctly - snapshot', async () => {
        const { asFragment } = render(
            <>
                <FormSelectConditionFilter />
                <FormInputConditionFilter />
                <FormInputSingleConditionFilter />
            </>
        )

        await waitFor(() => {
            expect(asFragment()).toMatchSnapshot()
        })
    })

    it('toggle content', async () => {
        const { container } = render(
            <>
                <FormSelectConditionFilter isTest dataTestId='condition-filter' />
            </>
        )

        const buttons = container.querySelectorAll('[data-test-id="condition-filter-header"]')
        forEach(buttons, (button) => {
            fireEvent.click(button)
        })

        await waitFor(() => {
            expect(container.querySelector('[data-test-id="condition-filter-content"]')).toBeInTheDocument()
            expect(container.querySelector('[data-test-id="condition-filter-content"]')).toBeVisible()
        })
    })
})
