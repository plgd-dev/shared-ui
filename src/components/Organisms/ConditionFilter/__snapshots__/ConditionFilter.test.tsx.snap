// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConditionFilter> render correctly - snapshot 1`] = `
<DocumentFragment>
  .emotion-0 {
  border-radius: 8px;
  border: 1px solid #E6E9ED;
  overflow: hidden;
}

.emotion-1 {
  padding: 16px 24px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  cursor: pointer;
  background: #FFFFFF;
}

.emotion-1:hover {
  background: #F6F7F9;
}

.emotion-1:hover .expander {
  color: #2261AE;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 8px;
}

.emotion-3 {
  font: 'Circular Pro',sans-serif;
  font-size: 16px;
  font-style: normal;
  font-weight: 700;
  line-height: 150%;
  letter-spacing: -0.5px;
  color: #0A2965;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 1px 10px;
  gap: 10px;
  border-radius: 44px;
  color: #81868C;
  background: #F6F7F9;
  font-family: 'Poppins',sans-serif;
  font-style: normal;
  font-weight: 400;
  font-size: 12px;
  line-height: 18px;
}

.emotion-5 {
  -webkit-flex: 0 0 32px;
  -ms-flex: 0 0 32px;
  flex: 0 0 32px;
}

.emotion-6 {
  color: #191A1A;
  background: #F6F7F9;
  padding: 6px;
  border-radius: 8px;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  width: 32px;
  height: 32px;
}

<div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      <div
        class="emotion-2"
      >
        <div
          class="emotion-3"
        >
          FormSelect
        </div>
        <div
          class="emotion-4"
        >
          Not set
        </div>
      </div>
      <div
        class="emotion-5"
      >
        <div
          class="expander emotion-6"
        >
          <svg
            fill="none"
            height="20"
            viewBox="0 0 20 20"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.59 6.912a.833.833 0 1 0-1.18 1.179l5 5a.833.833 0 0 0 1.18 0l5-5a.833.833 0 0 0-1.18-1.179L10 11.322z"
              fill="currentcolor"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
  .emotion-0 {
  border-radius: 8px;
  border: 1px solid #E6E9ED;
  overflow: hidden;
}

.emotion-1 {
  padding: 16px 24px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  cursor: pointer;
  background: #FFFFFF;
}

.emotion-1:hover {
  background: #F6F7F9;
}

.emotion-1:hover .expander {
  color: #2261AE;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 8px;
}

.emotion-3 {
  font: 'Circular Pro',sans-serif;
  font-size: 16px;
  font-style: normal;
  font-weight: 700;
  line-height: 150%;
  letter-spacing: -0.5px;
  color: #0A2965;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 1px 10px;
  gap: 10px;
  border-radius: 44px;
  color: #81868C;
  background: #F6F7F9;
  font-family: 'Poppins',sans-serif;
  font-style: normal;
  font-weight: 400;
  font-size: 12px;
  line-height: 18px;
}

.emotion-5 {
  -webkit-flex: 0 0 32px;
  -ms-flex: 0 0 32px;
  flex: 0 0 32px;
}

.emotion-6 {
  color: #191A1A;
  background: #F6F7F9;
  padding: 6px;
  border-radius: 8px;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  width: 32px;
  height: 32px;
}

<div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      <div
        class="emotion-2"
      >
        <div
          class="emotion-3"
        >
          FormInput
        </div>
        <div
          class="emotion-4"
        >
          Not set
        </div>
      </div>
      <div
        class="emotion-5"
      >
        <div
          class="expander emotion-6"
        >
          <svg
            fill="none"
            height="20"
            viewBox="0 0 20 20"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.59 6.912a.833.833 0 1 0-1.18 1.179l5 5a.833.833 0 0 0 1.18 0l5-5a.833.833 0 0 0-1.18-1.179L10 11.322z"
              fill="currentcolor"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
  .emotion-0 {
  border-radius: 8px;
  border: 1px solid #E6E9ED;
  overflow: hidden;
}

.emotion-1 {
  padding: 16px 24px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  cursor: pointer;
  background: #FFFFFF;
}

.emotion-1:hover {
  background: #F6F7F9;
}

.emotion-1:hover .expander {
  color: #2261AE;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 8px;
}

.emotion-3 {
  font: 'Circular Pro',sans-serif;
  font-size: 16px;
  font-style: normal;
  font-weight: 700;
  line-height: 150%;
  letter-spacing: -0.5px;
  color: #0A2965;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 1px 10px;
  gap: 10px;
  border-radius: 44px;
  color: #81868C;
  background: #F6F7F9;
  font-family: 'Poppins',sans-serif;
  font-style: normal;
  font-weight: 400;
  font-size: 12px;
  line-height: 18px;
}

.emotion-5 {
  -webkit-flex: 0 0 32px;
  -ms-flex: 0 0 32px;
  flex: 0 0 32px;
}

.emotion-6 {
  color: #191A1A;
  background: #F6F7F9;
  padding: 6px;
  border-radius: 8px;
  -webkit-transition: all 0.3s;
  transition: all 0.3s;
  width: 32px;
  height: 32px;
}

<div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      <div
        class="emotion-2"
      >
        <div
          class="emotion-3"
        >
          FormInput single
        </div>
        <div
          class="emotion-4"
        >
          Not set
        </div>
      </div>
      <div
        class="emotion-5"
      >
        <div
          class="expander emotion-6"
        >
          <svg
            fill="none"
            height="20"
            viewBox="0 0 20 20"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.59 6.912a.833.833 0 1 0-1.18 1.179l5 5a.833.833 0 0 0 1.18 0l5-5a.833.833 0 0 0-1.18-1.179L10 11.322z"
              fill="currentcolor"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
